src/cswinrt/code_writers.h

```diff
@@ -3549,6 +3550,473 @@ private % AsInternal(InterfaceTag<%> _) => % ?? Make_%();
+        auto projection_name = w.write_temp("%", bind<write_projection_type>(type));
+        auto abi_name = w.write_temp("%", bind<write_abi_type>(type));
+        w.write(
+R"(internal sealed unsafe class %ComWrappersMarshallerAttribute : WindowsRuntimeComWrappersMarshallerAttribute
+{
+    public override void* GetOrCreateComInterfaceForObject(object value)
+    {
+        return WindowsRuntimeValueTypeMarshaller.BoxToUnmanaged<%>((%) value, in %ReferenceImpl.IID_IReferenceOf%).DetachThisPtrUnsafe();
+    }
+
+    public override ComInterfaceEntry* ComputeVtables(out int count)
+    {
+        count = sizeof(%InterfaceEntries) / sizeof(ComInterfaceEntry);
+        return (ComInterfaceEntry*)Unsafe.AsPointer(in %InterfaceEntriesImpl.Entries);
+    }
+
+    public override object CreateObject(void* value, out CreatedWrapperFlags wrapperFlags)
+    {
+        wrapperFlags = CreatedWrapperFlags.NonWrapping;
+        return WindowsRuntimeValueTypeMarshaller.UnboxToManagedUnsafe<%>(value, in %ReferenceImpl.IID_IReferenceOf%);
+    }
+}
+)", name, projection_name, projection_name, name, name, name, name, is_type_blittable(type) ? projection_name : abi_name, name, name);
+    }
+
+    void write_interface_entries_impl(writer& w, TypeDef const& type)
+    {
+        auto name = w.write_temp("%", bind<write_type_name>(type, typedef_name_type::ABI, false));
+
+        w.write(
+R"(file static class %InterfaceEntriesImpl
+{
+    [FixedAddressValueType]
+    public static readonly %InterfaceEntries Entries;
+    
+    static %InterfaceEntriesImpl()
+    {
+        Entries.IReferenceOf%.IID = %ReferenceImpl.IID_IReferenceOf%;
+        Entries.IReferenceOf%.Vtable = %ReferenceImpl.Vtable;
+        Entries.IPropertyValue.IID = IPropertyValueImpl.IID;
+        Entries.IPropertyValue.Vtable = IPropertyValueImpl.OtherTypeVtable;
+        Entries.IStringable.IID = IStringableImpl.IID;
+        Entries.IStringable.Vtable = IStringableImpl.Vtable;
+        Entries.IWeakReferenceSource.IID = IWeakReferenceSourceImpl.IID;
+        Entries.IWeakReferenceSource.Vtable = IWeakReferenceSourceImpl.Vtable;
+        Entries.IMarshal.IID = IMarshalImpl.IID;
+        Entries.IMarshal.Vtable = IMarshalImpl.Vtable;
+        Entries.IAgileObject.IID = IAgileObjectImpl.IID;
+        Entries.IAgileObject.Vtable = IAgileObjectImpl.Vtable;
+        Entries.IInspectable.IID = IInspectableImpl.IID;
+        Entries.IInspectable.Vtable = IInspectableImpl.Vtable;
+        Entries.IUnknown.IID = IUnknownImpl.IID;
+        Entries.IUnknown.Vtable = IUnknownImpl.Vtable;
+    }
+}
+)", name, name, name, name, name, name, name, name);
+    }
+
+    void write_com_interface_entries(writer& w, TypeDef const& type)
+    {
+        auto name = w.write_temp("%", bind<write_typedef_name>(type, typedef_name_type::ABI, false));
+
+        w.write(
+R"([StructLayout(LayoutKind.Sequential)]
+file struct %InterfaceEntries
```

Comment:
I think you should be able to use a single copy of this type as well.

Reply:
Done